plugins/node/instrumentation-tedious/src/instrumentation.ts

```diff
@@ -0,0 +1,226 @@
+        this.removeListener('databaseChange', setDatabase);
+      });
+      return original.apply(this, arguments as unknown as any[]);
+    };
+  }
+
+  private _patchQuery(operation: string) {
+    return (originalMethod: UnknownFunction): UnknownFunction => {
+      const thisPlugin = this;
+      this._diag.debug(
+        `TediousInstrumentation: patched Connection.prototype.${operation}`
+      );
+
+      function patchedMethod(this: ApproxConnection, request: ApproxRequest) {
+        let procCount = 0;
+        let statementCount = 0;
+        const incrementStatementCount = () => statementCount++;
+        const incrementProcCount = () => procCount++;
+        const databaseName = this[CURRENT_DATABASE];
+        const sql = (request => {
+          // Required for <11.0.9
+          if (
+            request.sqlTextOrProcedure === 'sp_prepare' &&
+            request?.parametersByName?.stmt?.value
+          ) {
+            return request.parametersByName.stmt.value;
+          }
+          return request.sqlTextOrProcedure;
+        })(request);
+
+        const span = thisPlugin.tracer.startSpan(
+          getSpanName(operation, databaseName, sql, request.table),
+          {
+            kind: api.SpanKind.CLIENT,
+            attributes: {
+              [SemanticAttributes.DB_SYSTEM]: DbSystemValues.MSSQL,
+              [SemanticAttributes.DB_NAME]: databaseName,
+              [SemanticAttributes.NET_PEER_PORT]: this.config?.options?.port,
+              [SemanticAttributes.NET_PEER_NAME]: this.config?.server,
+              // >=4 uses `authentication` object, older versions just userName and password pair
+              [SemanticAttributes.DB_USER]:
+                this.config.userName ??
+                this.config?.authentication?.options?.userName,
+              [SemanticAttributes.DB_STATEMENT]: sql,
+              [SemanticAttributes.DB_SQL_TABLE]: request.table,
+            },
+          }
+        );
+
+        const endSpan = once((err?: any) => {
+          request.removeListener('done', incrementStatementCount);
+          request.removeListener('doneInProc', incrementStatementCount);
+          request.removeListener('doneProc', incrementProcCount);
+          request.removeListener('error', endSpan);
+          this.removeListener('end', endSpan);
+
+          span.setAttribute('tedious.procedure_count', procCount);
+          span.setAttribute('tedious.statement_count', statementCount);
+          if (err) {
+            span.setStatus({
+              code: api.SpanStatusCode.ERROR,
+              message: err.message,
+            });
+          }
+          span.end();
+        });
+
+        request.on('done', incrementStatementCount);
+        request.on('doneInProc', incrementStatementCount);
+        request.on('doneProc', incrementProcCount);
+        request.once('error', endSpan);
+        this.on('end', endSpan);
+
+        if (typeof request.callback === 'function') {
```

Comment:
Actually, this check here is just a safety net. `callback` is not set by the user but an internal one. There is another property called `userCallback` on Request instance for the user-provided callback.

We expect the callback to always exist as a function.

Reply:
Cool,
How do you feel about writing a log error in this case? as effectively, when this `if` is not hit, the instrumentation will not end most spans and will not remove the request listeners at `endSpan`

Reply:
We can still rely on `end` or `error` events to be fired.

Alternatively, we could just synchronously end span after calling the original method. I think that would be covering cases we don't know about or think are not possible, but otherwise of little benefit. WDYT?

I added a log statement to at least comment on the case.

Reply:
If we expect `callback` to always be set, and we have a log message when that fails, then I think it's good enough.
Thanks for addressing this :)